app/src/main/cpp/TrackedKeyboardRenderer.cpp

```diff
@@ -0,0 +1,440 @@
+    modelLoader.SetImageLoader(&TrackedKeyboardRenderer::ImageLoaderCallback, &ktxTex);
+
+    if (!modelLoader.LoadBinaryFromMemory(&model, &err, &warn, modelBuffer.data(), modelBuffer.size())) {
+        VRB_ERROR("XR_FB_keyboard_tracking: Error loading keyboard model: %s", err.c_str());
+        return false;
+    }
+
+    if (!warn.empty())
+        VRB_WARN("%s", warn.c_str());
+
+    if (model.meshes.size() == 0)
+        return false;
+    // Assume the first mesh
+    auto& mesh = model.meshes[0];
+
+    if (mesh.primitives.size() == 0)
+        return false;
+    // Assume the first primitive of the mesh
+    auto& primitive = mesh.primitives[0];
+
+    // Load indices
+    if (primitive.indices < 0 || primitive.indices >= model.accessors.size())
+        return false;
+    gltfModel.indicesAttrIndex = primitive.indices;
+
+    // Load vertex attributes
+    for (auto& attr: primitive.attributes) {
+        if (attr.second >= model.accessors.size())
+            return false;
+
+        if (attr.first == "POSITION") {
+            gltfModel.positionAttrIndex = attr.second;
+        } else if (attr.first == "NORMAL") {
+            gltfModel.normalAttrIndex = attr.second;
+        } else if (attr.first == "TEXCOORD_0") {
+            gltfModel.texcoordAttrIndex = attr.second;
+        }
+    }
+
+    if (gltfModel.positionAttrIndex == -1 || gltfModel.normalAttrIndex == -1 || gltfModel.texcoordAttrIndex == -1)
+        return false;
+
+    m.model = std::make_unique<GLTFModel>(gltfModel);
+    m.model->ktxTex = ktxTex;
+
+    VRB_LOG("XR_FB_keyboard_tracking: Keyboard model loaded successfully");
+
+    return true;
+}
+
+void TrackedKeyboardRenderer::SetVisible(const bool aVisible) {
+    m.visible = aVisible;
+}
+
+bool TrackedKeyboardRenderer::InitializeGL() {
+    m.vertexShader = vrb::LoadShader(GL_VERTEX_SHADER, sVertexShader);
+    m.fragmentShader = vrb::LoadShader(GL_FRAGMENT_SHADER, sFragmentShader);
+    if (m.vertexShader && m.fragmentShader)
+        m.program = vrb::CreateProgram(m.vertexShader, m.fragmentShader);
+
+    if (!m.program) {
+        VRB_ERROR("XR_FB_keyboard_tracking: Error creating GL program");
+        return false;
+    }
+
+    m.aPosition = vrb::GetAttributeLocation(m.program, "a_position");
+    m.aNormal = vrb::GetAttributeLocation(m.program, "a_normal");
+    m.aTexcoord = vrb::GetAttributeLocation(m.program, "a_texcoord");
+
+    m.uPerspective = vrb::GetUniformLocation(m.program, "u_perspective");
+    m.uView = vrb::GetUniformLocation(m.program, "u_view");
+    m.uModel = vrb::GetUniformLocation(m.program, "u_model");
+    m.uSampler = vrb::GetUniformLocation(m.program, "u_sampler");
+    VRB_GL_CHECK(glUniform1i(m.uSampler, 0));
+
+    glFlush();
+
+    return true;
+}
+
+void TrackedKeyboardRenderer::Draw(const vrb::Camera& aCamera) {
+    if (!m.visible || m.model == nullptr)
+        return;
+
+    if (!m.program)
+        InitializeGL();
+
+    assert(m.program);
```

Comment:
A valid code path in InitializeGL() returns false and does not create the program (if any of the shaders is not loaded). However here we have an assert. So I'd say we either consider that code path as valid and thus remove this assert and return, or we become stricted in the `InitializeGL()` method. The current implementation seems a bit inconsistent to me, on the one hand we consider a code path as valid (although erroneous) while in the other we assert, hope I have explained myself properly.